rfcs/0000-ret-cont-recursive-nix.md

```diff
@@ -0,0 +1,245 @@
+build(drv0) = { succeeded = build0; }
+isTrustlessStore(build0_store)
+drv1 = read(build0_store + drv1path)
+readlink(build0_drv) = build0.store + drv1path
+-------------------------------------------------- immediate-drv-deligation
+reducesTo(drv0, drv1)
+```
+```
+drv0, drv1, drv2 : Drv
+reducesTo(drv1, drv2)
+immediatelyDependsOn(drv0, drv1)
+-------------------------------------------------- transitive-drv-deligation
+reducesTo(drv0, drv0[drv2/drv1])
+```
+```
+drv0, drv1 : Drv
+reducesTo(drv0, drv1)
+∀<o : outputs(drv0)> build0_o : StorePath
+∀<o : outputs(drv0)> assoc(drv0, o, build0_o)
+-------------------------------------------------- delegative-build
+∀<o : outputs(drv0)> assoc(drv1, o, build1_o)
+```
+
+## Design Notes
+
+There's a few things we can call out from the faux-formalization.
+
+ - `isTrustlessStore` is called that because the restricted on the contents—fixed output builds / plain data and drvs—is fully and cheaply verifiabled.
+   This is in contrast to normal builds,
+   where the relationship between the derivation and build can only be verified by redoing the build,
+   and where even then there's no way to know whether to blame the output for being actually malicious, or the derivation for merely being non-deterministic.
+
+ - The substitution of drvs in a downstream derivation reminds me of the substitution of drvs for content hashes with the intensional store.
+   We should muse on this point, and hopefully write a small-step semantics for both together that is more elegant than the above.
+
+ - The *building* itself of derivations is unchanged.
+   All the magic happens through the `reducesTo` relation.
+
+ - Because drvs can produce plans of drvs producing more drvs ad-infinitum, it's possible to never terminate (no `reducesTo` from a derivation to an `isReadyToBuild` derivation) but that's the user's fault.
+   We can detect simple cycles analogous to black holes in thunks: if a derivation produces a redirected derivation depending on the original, a cycle is effectively recreated even though we don't have a hash fixed point.
+   Nix should raise an error rather than looping, but either behavior is permissible.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+ - The opinionated nature may put off those who think Nix is too hard to learn already, and think simple recursive "nix-build" is good for newcomers.
+
+ - If we ever want full recursive Nix, this doesn't really build in that direction.
+   It sidesteps the bulk of the difficulty which is in making the nested sandboxing and daemon communication secure.
+   To me though, this is a feature not a bug; I don't want to go in that direction just yet.
+
+# Alternatives
+[alternatives]: #alternatives
+
+ - Don't allow fixed-output builds.
+   All data can be stuck inside the drv file, so this can be cut without limiting expressive power.
+   But this is much less efficient, and more cumbersome for whatever produces the data.
+
+ - Use a socket to talk to the host daemon.
```

Comment:
I'm *hugely* in favor of this approach. Using nested stores and all that is kind of hacky, and we have lots of reasons besides actually nested builds to want recursive nix (e.g. store querying, dynamic dependency fetching, etc.). In general I'm in favor of moving toward a much more structured way for builds that are nix-aware to talk to the store; builds that *aren't* can then use generic functionality built on top of that.

Reply:
Yeah I don't really care soooo much on this point. I guess I am more interested in pointing out that we *can* skip the daemon socket, as proof of the simpler computational model that is this version, than insisting that we *must* skip the daemon socket.